crates/ruff_python_ast/src/operator_precedence.rs

```diff
@@ -0,0 +1,176 @@
+
+            // Unary operators: +x, -x, ~x (except boolean not)
+            ExprRef::UnaryOp(operator) => match operator.op {
+                UnaryOp::UAdd | UnaryOp::USub | UnaryOp::Invert => Self::PosNegBitNot,
+                UnaryOp::Not => Self::Not,
+            },
+
+            // Math binary ops
+            ExprRef::BinOp(binary_operation) => Self::from(binary_operation.op),
+
+            // Comparisons: <, <=, >, >=, ==, !=, in, not in, is, is not
+            ExprRef::Compare(_) => Self::ComparisonsMembershipIdentity,
+
+            // Boolean not
+            // Handled above in unary operators
+
+            // Boolean operations: and, or
+            ExprRef::BoolOp(bool_op) => Self::from(bool_op.op),
+
+            // Conditional expressions: x if y else z
+            ExprRef::If(_) => Self::IfElse,
+
+            // Lambda expressions
+            ExprRef::Lambda(_) => Self::Lambda,
+
+            // Unpacking also omitted in the docs, but has almost the lowest precedence,
+            // except for assignment & yield expressions. E.g. `[*(v := [1,2])]` is valid
+            // but `[*v := [1,2]] would fail on incorrect syntax because * will associate
+            // `v` before the assignment.
+            ExprRef::Starred(_) => Self::Starred,
+
+            // Assignment expressions (aka named)
+            ExprRef::Named(_) => Self::Assign,
+
+            // Although omitted in docs, yield expressions may be used inside an expression
+            // but must be parenthesized. So for our purposes we assume they just have
+            // the lowest "real" precedence.
+            ExprRef::Yield(_) | ExprRef::YieldFrom(_) => Self::Yield,
+
+            // Not a real python expression, so treat as lowest as well
+            ExprRef::IpyEscapeCommand(_) => Self::None,
+        }
+    }
+
+    pub fn from_expr(expr: &Expr) -> Self {
+        Self::from(&ExprRef::from(expr))
+    }
+}
+
+impl From<&Expr> for OperatorPrecedence {
+    fn from(expr: &Expr) -> Self {
+        Self::from_expr(expr)
+    }
+}
+
+impl<'a> From<&ExprRef<'a>> for OperatorPrecedence {
+    fn from(expr_ref: &ExprRef<'a>) -> Self {
+        Self::from_expr_ref(expr_ref)
+    }
+}
+
+impl From<Operator> for OperatorPrecedence {
+    fn from(operator: Operator) -> Self {
+        match operator {
+            // Multiplication, matrix multiplication, division, floor division, remainder:
+            // *, @, /, //, %
+            Operator::Mult
+            | Operator::MatMult
+            | Operator::Div
+            | Operator::Mod
+            | Operator::FloorDiv => Self::MulDivRemain,
+            // Addition, subtraction
+            Operator::Add | Operator::Sub => Self::AddSub,
+            // Bitwise shifts: <<, >>
+            Operator::LShift | Operator::RShift => Self::LeftRightShift,
+            // Bitwise operations: &, ^, |
+            Operator::BitAnd => Self::BitAnd,
+            Operator::BitXor | Operator::BitOr => Self::BitXorOr,
```

Comment:
I don't think both `BitXor` and `BitOr` are at the same precedence level (https://docs.python.org/3/reference/expressions.html#operator-precedence).